src/hotspot/share/cds/classPrelinker.cpp

```diff
@@ -0,0 +1,224 @@
+  _singleton = NULL;
+}
+
+bool ClassPrelinker::can_archive_resolved_vm_class(InstanceKlass* cp_holder, InstanceKlass* resolved_klass) {
+  if (!is_vm_class(resolved_klass)) {
+    return false;
+  }
+  if (!cp_holder->is_shared_boot_class() &&
+      !cp_holder->is_shared_platform_class() &&
+      !cp_holder->is_shared_app_class()) {
+    // Custom loaders are not guaranteed to resolve the vmClasses to the
+    // ones resolved by the boot loader.
+    return false;
+  }
+  if (cp_holder->class_loader_data() != resolved_klass->class_loader_data()) {
+    // If they are defined by different loaders, it's possible for resolved_klass
+    // to be already defined, but is not yet resolved in cp_holder->class_loader().
+
+    // TODO: this check can be removed if we preload the vmClasses into
+    // platform and app loaders during VM bootstrap.
+    return false;
+  }
+
+  return true;
+}
+
+bool ClassPrelinker::can_archive_resolved_klass(InstanceKlass* cp_holder, Klass* resolved_klass) {
+  assert(!is_in_archivebuilder_buffer(cp_holder), "sanity");
+  assert(!is_in_archivebuilder_buffer(resolved_klass), "sanity");
+
+  if (resolved_klass->is_instance_klass()) {
+    InstanceKlass* ik = InstanceKlass::cast(resolved_klass);
+    if (can_archive_resolved_vm_class(cp_holder, ik)) {
+      return true;
+    }
+    if (cp_holder->is_subtype_of(ik)) {
+      // All super types of ik will be resolved in ik->class_loader() before
+      // ik is defined in this loader, so it's safe to archive the resolved klass reference.
+      return true;
+    }
+
+    // TODO -- allow objArray classes, too
+  }
+
+  return false;
+}
+
+Klass* ClassPrelinker::get_resolved_klass_or_null(ConstantPool* cp, int cp_index) {
+  if (cp->tag_at(cp_index).is_klass()) {
+    CPKlassSlot kslot = cp->klass_slot_at(cp_index);
+    int resolved_klass_index = kslot.resolved_klass_index();
+    return cp->resolved_klasses()->at(resolved_klass_index);
+  } else {
+    // klass is not resolved yet
+    assert(cp->tag_at(cp_index).is_unresolved_klass() ||
+           cp->tag_at(cp_index).is_unresolved_klass_in_error(), "must be");
+    return NULL;
+  }
+}
+
+bool ClassPrelinker::can_archive_resolved_klass(ConstantPool* cp, int cp_index) {
+  assert(!is_in_archivebuilder_buffer(cp), "sanity");
+  assert(cp->tag_at(cp_index).is_klass(), "must be resolved");
+
+  Klass* resolved_klass = get_resolved_klass_or_null(cp, cp_index);
+  assert(resolved_klass != NULL, "must be");
+
+  return can_archive_resolved_klass(cp->pool_holder(), resolved_klass);
+}
+
+void ClassPrelinker::dumptime_resolve_constants(InstanceKlass* ik, TRAPS) {
+  constantPoolHandle cp(THREAD, ik->constants());
+  if (cp->cache() == NULL || cp->reference_map() == NULL) {
+    // The cache may be NULL if the pool_holder klass fails verification
+    // at dump time due to missing dependencies.
+    return;
+  }
+
+  bool first_time;
+  _processed_classes.put_if_absent(ik, &first_time);
```

Comment:
I don't see any get functions for this hashtable?

Reply:
This table is used to check if we have already worked on the class:

```
  bool first_time;
  _processed_classes.put_if_absent(ik, &first_time);
  if (!first_time) {
    return;
  }
```

Reply:
Oh, ok I missed that.  Can you add a comment there about what you're doing and why?

Reply:
Done